op-node/sources/sync_client.go

```diff
@@ -51,41 +65,87 @@ func NewSyncClient(receiver receivePayload, client client.RPC, log log.Logger, m
 	if err != nil {
 		return nil, err
 	}
-
+	// This resource context is shared between all workers that may be started
+	resCtx, resCancel := context.WithCancel(context.Background())
 	return &SyncClient{
-		L2Client:         l2Client,
-		FetchUnsafeBlock: make(chan uint64, 128),
-		done:             make(chan struct{}),
-		receivePayload:   receiver,
+		L2Client:       l2Client,
+		resCtx:         resCtx,
+		resCancel:      resCancel,
+		requests:       make(chan syncRequest, 128),
+		receivePayload: receiver,
 	}, nil
 }
 
-// Start starts up the state loop.
-// The loop will have been started if err is not nil.
+// Start starts the syncing background work. This may not be called after Close().
 func (s *SyncClient) Start() error {
+	// TODO(CLI-3635): we can start multiple event loop runners as workers, to parallelize the work
 	s.wg.Add(1)
 	go s.eventLoop()
 	return nil
 }
 
-// Close sends a signal to the event loop to stop.
+// Close sends a signal to close all concurrent syncing work.
 func (s *SyncClient) Close() error {
-	s.done <- struct{}{}
+	s.resCancel()
 	s.wg.Wait()
 	return nil
 }
 
+func (s *SyncClient) RequestL2Range(ctx context.Context, start, end uint64) error {
+	// Drain previous requests now that we have new information
+	for len(s.requests) > 0 {
+		<-s.requests
+	}
+
+	// TODO(CLI-3635): optimize the by-range fetching with the Engine API payloads-by-range method.
+
+	s.log.Info("Scheduling to fetch missing payloads from backup RPC", "start", start, "end", end, "size", end-start)
+
+	for i := start; i < end; i += maxRangePerWorker {
+		r := syncRequest{start: i, end: i + maxRangePerWorker}
+		if r.end > end {
+			r.end = end
+		}
+		s.log.Info("Scheduling range request", "start", r.start, "end", r.end, "size", r.end-r.start)
+		// schedule new range to be requested
+		select {
+		case s.requests <- r:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
 // eventLoop is the main event loop for the sync client.
 func (s *SyncClient) eventLoop() {
 	defer s.wg.Done()
 	s.log.Info("Starting sync client event loop")
 
 	for {
 		select {
-		case <-s.done:
+		case <-s.resCtx.Done():
+			s.log.Debug("Shutting down RPC sync worker")
 			return
-		case blockNumber := <-s.FetchUnsafeBlock:
-			s.fetchUnsafeBlockFromRpc(context.Background(), blockNumber)
+		case r := <-s.requests:
+			// Limit the maximum time for fetching payloads
+			ctx, cancel := context.WithTimeout(s.resCtx, time.Second*10)
+			// We are only fetching one block at a time here.
+			err := s.fetchUnsafeBlockFromRpc(ctx, r.start)
+			cancel()
+			if err != nil {
+				s.log.Error("failed syncing L2 block via RPC", "err", err)
+			} else {
+				r.start += 1 // continue with next block
+			}
+			// Reschedule
+			if r.start < r.end {
+				select {
+				case s.requests <- r:
+				default:
+					// drop syncing job if we are too busy with sync jobs already.
+				}
+			}
```

Comment:
Does this mean the request is rescheduled at the end of the channel? So we'd break the range to request up into blocks of 10, then request the first block from each, then the second from each etc. That works, but seems rather inefficient.  Any reason we don't just have a for loop in here that requests all the blocks in the batch one at a time?

Reply:
The requests in the channel are meant to be a small range, see p2p sync PR which uses a similar mechanism but chops it up like that. Here we currently only really have one item in the channel (we drain on new information), don't do the splitting yet, and don't sync from multiple RPC sources (yet, would like to change the sync-client to a list of sync clients and have some parallel fetching + redundancy that way).

Reply:
Up on line 105 we're splitting the single incoming request into batches of 10 before we push them into the `s.requests` channel:

```go
	for i := start; i < end; i += maxRangePerWorker {
 		r := syncRequest{start: i, end: i + maxRangePerWorker}
```

Unless I'm really misunderstanding the flow here...

Reply:
This is still requesting the first block of each range in turn rather than requesting block sequentially.  Did we deliberately decide to stick with this or did a change get lost here too?

Reply:
Pro: reduce range by 1 item allows us to remove the request and update fully, avoiding consecutive requests for the same L2 block number to be in the channel (although not ruling out that an in-flight request isn't duplicate)

Con: the batches result in out of order requests if the channel is filled with multiple requests.

Adding a small cache to handle duplicate requests, and doing 1-by-1 probably works better, at least while we don't support by-range RPC requests. Will change it.

Reply:
Opted to change it back to by-block-number request scheduling, but now with re-attempts + exponential backoff on the request. If it continues to fail, it will reschedule at the end of the requests queue. And once there's a new range, it'll still drop all previously scheduled work and schedule the new range.